File: android/app/src/test/java/mullu/comrade/ui/ChatCommandsTest.kt
package mullu.comrade.ui

import mullu.comrade.ComradeCore
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import uniffi.comrade_core.AppAction
import uniffi.comrade_core.ChatCommand
import uniffi.comrade_core.CommandSpec
import uniffi.comrade_core.Mention
import uniffi.comrade_core.MusicService
import uniffi.comrade_ui.PlayRoute

/**
 * Mirrors `desktop/ui/chat_commands.test.mjs` case for case, so the two
 * frontends cannot drift on what a parsed command means. Where they differ on
 * purpose — Android has all five screens, desktop has two — each side's test
 * says so rather than one of them being silently wrong.
 */
class ChatCommandsTest {

    private fun mention(handle: String) =
        Mention(handle = handle, start = 0u, end = (handle.length + 1).toUInt())

    private fun resolved(handle: String, npub: String) = ComradeCore.MentionMatchInfo(
        handle = handle,
        start = 0,
        end = handle.length + 1,
        npub = npub,
        candidates = emptyList(),
    )

    private fun unknown(handle: String) = ComradeCore.MentionMatchInfo(
        handle = handle,
        start = 0,
        end = handle.length + 1,
        npub = null,
        candidates = emptyList(),
    )

    private fun ambiguous(handle: String) = ComradeCore.MentionMatchInfo(
        handle = handle,
        start = 0,
        end = handle.length + 1,
        npub = null,
        candidates = listOf(
            ComradeCore.ContactInfo("npub1a", "ana", null, false),
            ComradeCore.ContactInfo("npub1b", "ana", null, false),
        ),
    )

    private fun spec(
        name: String,
        aliases: List<String> = emptyList(),
        argument: String = "",
        takesMention: Boolean = false,
    ) = CommandSpec(
        name = name,
        aliases = aliases,
        argument = argument,
        help = "help for $name",
        takesMention = takesMention,
    )

    @Test
    fun ordinaryTextIsSentAsTyped() {
        assertEquals(ComposerPlan.Send, ChatCommands.planFor(ChatCommand.Plain))
    }

    @Test
    fun payStillGoesOutAsText() {
        // Recognised so it is never called a typo; nothing else changes.
        assertEquals(ComposerPlan.Send, ChatCommands.planFor(ChatCommand.Pay))
    }

    @Test
    fun anUnknownCommandIsReportedRatherThanSwallowed() {
        // A message silently eaten is worse than one that arrives confusing —
        // the sender believes they sent it.
        val plan = ChatCommands.planFor(ChatCommand.Unknown("frobnicate"))
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("/frobnicate"))
    }

    // ── Tara asides ──────────────────────────────────────────────────────────

    @Test
    fun anAsideRoutesToTaraNotToThePeer() {
        val plan = ChatCommands.planFor(ChatCommand.AskTara("i keep putting this off"))
        assertEquals(ComposerPlan.Aside("i keep putting this off"), plan)
    }

    @Test
    fun anAddressedButEmptyAsidePromptsInsteadOfSendingNothing() {
        val plan = ChatCommands.planFor(ChatCommand.AskTara("  "))
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("only you will see it"))
    }

    @Test
    fun theComposerKnowsTheAudienceFromTheMomentTaraIsAddressed() {
        // Eager on purpose, and now in two directions: `/tara` never leaves the
        // device and `@tara` reaches the other person. Either label being wrong
        // is how somebody publishes a private thought, or asks a question the
        // peer turns out never to have seen.
        assertEquals(ChatCommands.TaraAudience.Shared, ChatCommands.taraDraft("@tara "))
        assertEquals(ChatCommands.TaraAudience.Shared, ChatCommands.taraDraft("@tara"))
        assertEquals(ChatCommands.TaraAudience.Shared, ChatCommands.taraDraft("@Tara what about"))
        assertEquals(ChatCommands.TaraAudience.Shared, ChatCommands.taraDraft("  @tara hello"))
        assertEquals(ChatCommands.TaraAudience.Private, ChatCommands.taraDraft("/tara hello"))
        assertEquals(ChatCommands.TaraAudience.Private, ChatCommands.taraDraft("/tara"))
    }

    @Test
    fun aPersonWhoseHandleStartsWithTaraIsNotTara() {
        assertEquals(null, ChatCommands.taraDraft("@taranjeet are you around"))
        assertEquals(null, ChatCommands.taraDraft("hello @tara"))
        assertEquals(null, ChatCommands.taraDraft(""))
    }

    // ── Tara in the room (`@tara …`) ──────────────────────────────────────────

    @Test
    fun sharedAndPrivateTaraAreDifferentPlans() {
        // Same words, two audiences. One plan reaches `tara_in_chat` (two
        // messages into the thread), the other `tara_aside` (nothing sent).
        assertEquals(
            ComposerPlan.TaraHere("what film should we watch"),
            ChatCommands.planFor(ChatCommand.TaraHere("what film should we watch")),
        )
        assertEquals(
            ComposerPlan.Aside("what film should we watch"),
            ChatCommands.planFor(ChatCommand.AskTara("what film should we watch")),
        )
    }

    @Test
    fun anEmptySharedAddressSaysTheAnswerWillBeSeenByBoth() {
        val plan = ChatCommands.planFor(ChatCommand.TaraHere(" "))
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("you'll both see"))
    }

    @Test
    fun theHelplinesAreShownAsLinesWhereThereIsNoCardToShowThem() {
        // AUDIT §8's honesty gate: the resources go beside *any* distress reply,
        // and the composer's note is a single Text with no card slot.
        val lines = ChatCommands.crisisLines(
            listOf(
                ComradeCore.CrisisResourceInfo("Tele-MANAS (India)", "14416", "24×7"),
                ComradeCore.CrisisResourceInfo("KIRAN", "1800-599-0019", "24×7"),
            ),
        )
        assertTrue(lines.contains("14416"))
        assertTrue(lines.contains("KIRAN"))
        // Empty in, empty out — a note must not end on a dangling heading.
        assertEquals("", ChatCommands.crisisLines(emptyList()))
    }

    // ── Tasks ────────────────────────────────────────────────────────────────

    @Test
    fun aTaskWithNobodyNamedIsANoteToSelf() {
        val plan = ChatCommands.planFor(ChatCommand.Task("water the plants", emptyList()))
        assertEquals(ComposerPlan.Task("water the plants", null), plan)
    }

    @Test
    fun aTaskNamingAKnownContactCarriesTheirNpub() {
        val plan = ChatCommands.planFor(
            ChatCommand.Task("get some work done", listOf(mention("ana"))),
            listOf(resolved("ana", "npub1ana")),
        )
        assertEquals(ComposerPlan.Task("get some work done", "npub1ana"), plan)
    }

    @Test
    fun anEmptyTaskAsksWhatNeedsDoing() {
        val plan = ChatCommands.planFor(ChatCommand.Task("   ", emptyList()))
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("What needs doing"))
    }

    @Test
    fun aHandleThatIsNobodysIsReportedRatherThanGuessedAt() {
        val plan = ChatCommands.planFor(
            ChatCommand.Task("do it", listOf(mention("bina"))),
            listOf(unknown("bina")),
        )
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("@bina isn't in your contacts"))
    }

    @Test
    fun twoContactsAnsweringToOneHandleBecomeAQuestionNotACoinFlip() {
        // A handle is a self-declared alias; anyone may publish any name.
        // Picking one is how a private message reaches the wrong person.
        val plan = ChatCommands.planFor(
            ChatCommand.Task("do it", listOf(mention("ana"))),
            listOf(ambiguous("ana")),
        )
        assertTrue(plan is ComposerPlan.Choose)
        val choose = plan as ComposerPlan.Choose
        assertEquals("ana", choose.handle)
        assertEquals(2, choose.candidates.size)
    }

    @Test
    fun choosingOneOfTwoLetsTheSameDraftThrough() {
        // The fix for the dead end: the old plan said "pick which one" and gave
        // nothing to pick, so the command could never be completed at all.
        val mentions = ChatCommands.withChoices(
            listOf(ambiguous("ana")),
            mapOf("ana" to "npub1b"),
        )
        assertEquals(
            ComposerPlan.Task("do it", "npub1b"),
            ChatCommands.planFor(ChatCommand.Task("do it", listOf(mention("ana"))), mentions),
        )
    }

    @Test
    fun aChoiceThatNamesNobodyOnTheListIsIgnored() {
        // A pin left over from an earlier draft, or from before a contact was
        // removed, must not silently retarget a message — that is the exact
        // failure the ambiguity exists to prevent, arriving by another door.
        val stale = ChatCommands.withChoices(
            listOf(ambiguous("ana")),
            mapOf("ana" to "npub1someone-else"),
        )
        assertEquals(null, stale[0].npub)
        assertTrue(
            ChatCommands.planFor(
                ChatCommand.Task("do it", listOf(mention("ana"))),
                stale,
            ) is ComposerPlan.Choose,
        )
    }

    @Test
    fun aChooserRowCarriesTheKeyBecauseBothMayShareTheName() {
        // The usual reason a handle is ambiguous is that both people chose the
        // same name, so a row showing only the name offers two identical
        // choices — which is no choice at all.
        val a = ChatCommands.candidateLabel("ana", "npub1aaaaaaaaaaaaaaaaaaaa")
        val b = ChatCommands.candidateLabel("ana", "npub1bbbbbbbbbbbbbbbbbbbb")
        assertFalse(a == b)
        assertTrue(a.contains("ana"))
    }

    @Test
    fun anAmbiguousHandleInAnOfferAsksTooRatherThanPickingOne() {
        val plan = ChatCommands.planFor(
            ChatCommand.OfferTo(AppAction.BREATHE, listOf(mention("ana"))),
            listOf(ambiguous("ana")),
        )
        assertTrue(plan is ComposerPlan.Choose)
    }

    // ── Offers ───────────────────────────────────────────────────────────────

    @Test
    fun anOfferToAKnownComradeCarriesTheActionAndTheNpub() {
        val plan = ChatCommands.planFor(
            ChatCommand.OfferTo(AppAction.BREATHE, listOf(mention("ana"))),
            listOf(resolved("ana", "npub1ana")),
        )
        assertEquals(ComposerPlan.Offer(AppAction.BREATHE, listOf("npub1ana")), plan)
    }

    @Test
    fun anOfferWithNobodyNamedAsksWhoItIsFor() {
        val plan = ChatCommands.planFor(ChatCommand.OfferTo(AppAction.BREATHE, emptyList()))
        assertTrue(plan is ComposerPlan.Explain)
        assertTrue((plan as ComposerPlan.Explain).message.contains("Name the comrade"))
    }

    // ── Screens ──────────────────────────────────────────────────────────────

    @Test
    fun everyScreenThisAppHasOpens() {
        // Android has all five, which is the deliberate divergence from the
        // desktop module's two.
        for (action in AppAction.entries) {
            assertTrue("$action", action in ChatCommands.AVAILABLE_SCREENS)
            assertEquals(ComposerPlan.Open(action), ChatCommands.planFor(ChatCommand.Open(action)))
        }
    }

    @Test
    fun everyActionHasALabelForTheSentences() {
        for (action in AppAction.entries) {
            assertTrue("$action", ChatCommands.labelFor(action).isNotBlank())
        }
    }

    // ── Play ─────────────────────────────────────────────────────────────────

    @Test
    fun playCarriesTheQueryAndAnEmptyOneAsksForOne() {
        assertEquals(
            ComposerPlan.Play("Kun Faya Kun", null),
            ChatCommands.planFor(ChatCommand.Play("Kun Faya Kun", null)),
        )
        val empty = ChatCommands.planFor(ChatCommand.Play("  ", null))
        assertTrue(empty is ComposerPlan.Explain)
    }

    @Test
    fun playCarriesTheServiceTheCommandNamed() {
        // `/spotify kun faya kun` has to reach `play_query`, which is where a
        // link's own service is allowed to outrank the alias that was typed.
        assertEquals(
            ComposerPlan.Play("Kun Faya Kun", MusicService.SPOTIFY),
            ChatCommands.planFor(ChatCommand.Play("Kun Faya Kun", MusicService.SPOTIFY)),
        )
    }

    @Test
    fun everyPlayRouteSaysSomethingAndOnlyTheTwoThatPlayClaimTheyDid() {
        // Three of the six routes are refusals, and a refusal naming the wrong
        // reason sends someone looking for a file picker when the real problem
        // is that no account is connected. So each route gets its own sentence,
        // and none may be blank.
        val started = setOf(PlayRoute.START_TOGETHER, PlayRoute.PLAY_ON_SERVICE)
        for (route in PlayRoute.entries) {
            val note = ChatCommands.playNote(route, MusicService.SPOTIFY, "Kun Faya Kun")
            assertTrue("$route", note.isNotBlank())
            if (route !in started) {
                assertFalse("$route", note.contains("Playing"))
            }
        }
        for (route in started) {
            assertTrue(
                "$route",
                ChatCommands.playNote(route, MusicService.SPOTIFY, "Kun Faya Kun")
                    .contains("Kun Faya Kun"),
            )
        }
    }

    @Test
    fun theServiceRouteNamesWhereItIsPlayingAndTheRefusalNamesWhatIsMissing() {
        // The two Spotify sentences must not read alike: one is a live session
        // on this listener's own subscription, the other is "you have no account
        // here". Before 2026-08-08 the refusal blamed DRM, which was the wrong
        // reason — driving their client is exactly what their SDK is for.
        val playing = ChatCommands.playNote(PlayRoute.PLAY_ON_SERVICE, MusicService.SPOTIFY, "x")
        val refused = ChatCommands.playNote(PlayRoute.OPEN_ELSEWHERE, MusicService.SPOTIFY, "x")
        assertTrue(playing.contains("Spotify"))
        assertTrue(refused.contains("Spotify"))
        assertNotEquals(playing, refused)
    }

    @Test
    fun aServiceRefusalNamesTheServiceItCannotReach() {
        assertTrue(
            ChatCommands.playNote(PlayRoute.OPEN_ELSEWHERE, MusicService.SPOTIFY, "x")
                .contains("Spotify"),
        )
        assertTrue(
            ChatCommands.playNote(PlayRoute.OPEN_ELSEWHERE, MusicService.APPLE_MUSIC, "x")
                .contains("Apple Music"),
        )
        // No service on a DRM link is possible; a vague true sentence beats a
        // confidently wrong brand name.
        assertFalse(
            ChatCommands.playNote(PlayRoute.OPEN_ELSEWHERE, null, "x").contains("Spotify"),
        )
    }

    @Test
    fun everyServiceHasAName() {
        for (service in MusicService.entries) {
            assertTrue("$service", ChatCommands.serviceLabel(service).isNotBlank())
        }
        assertTrue(ChatCommands.serviceLabel(null).isNotBlank())
    }

    @Test
    fun notAllowedToLookReadsDifferentlyFromLookedAndNotThere() {
        // The two are different problems: one is answered by picking a file, the
        // other by granting access. A single sentence for both would send
        // someone hunting for a file that is already on the phone.
        val absent = ChatCommands.playNote(PlayRoute.ASK_FOR_FILE, null, "Kun Faya Kun")
        assertTrue(ChatCommands.LIBRARY_UNSEEN != absent)
        // It has to name *its* reason — being unable to look — because that is
        // the half the other sentence cannot say.
        assertTrue(ChatCommands.LIBRARY_UNSEEN.contains("library"))
        // And it must not claim absence, which is the thing it does not know.
        assertFalse(ChatCommands.LIBRARY_UNSEEN.contains("No copy"))
        // Neither sentence sends anyone to a screen any more: both routes open
        // the picker themselves (`ConversationScreen`'s `onPickTogetherFile`),
        // so naming a destination would be naming one already arrived at.
        for (note in listOf(ChatCommands.LIBRARY_UNSEEN, absent)) {
            assertFalse(note, note.contains("Open Together"))
        }
    }

    @Test
    fun theAskForAFileRouteDoesNotBlameTheService() {
        // "You have no copy" and "Spotify won't allow it" are different
        // problems with different next steps; conflating them is the failure
        // these two cases exist to prevent.
        val note = ChatCommands.playNote(PlayRoute.ASK_FOR_FILE, MusicService.SPOTIFY, "x")
        assertFalse(note.contains("Spotify"))
        // Still has to say what is being asked for, now that the picker opening
        // is the action and this sentence is only the reason it opened.
        assertTrue(note.contains("file"))
    }

    // ── The / picker ─────────────────────────────────────────────────────────

    private val catalog = listOf(
        spec("play", listOf("listen", "watch"), "<song>"),
        spec("task", listOf("todo"), "<what>", takesMention = true),
        spec("breathe", listOf("breath")),
    )

    @Test
    fun thePickerMatchesAliasesAsWellAsNames() {
        // An alias that works on submit but is invisible while typing reads as
        // not existing.
        val rows = ChatCommands.pickerRows("/lis", catalog)
        assertEquals(1, rows.size)
        assertEquals("play", rows[0].name)
    }

    @Test
    fun thePickerIsClosedForOrdinaryText() {
        assertTrue(ChatCommands.pickerRows("hello", catalog).isEmpty())
        assertTrue(ChatCommands.pickerRows("20/80 split", catalog).isEmpty())
        assertTrue(ChatCommands.pickerRows("/nomatch", catalog).isEmpty())
    }

    @Test
    fun thePickerClosesOnceTheCommandHasBeenChosen() {
        // A space means what follows is an argument, not a name.
        assertTrue(ChatCommands.pickerRows("/task ship it", catalog).isEmpty())
    }

    @Test
    fun aBareSlashOffersEverything() {
        assertEquals(3, ChatCommands.pickerRows("/", catalog).size)
    }

    @Test
    fun completionLeavesTheCaretWhereTheNextWordGoesOrSubmits() {
        assertEquals("/play ", ChatCommands.completionFor(catalog[0]))
        assertEquals("/task ", ChatCommands.completionFor(catalog[1]))
        // Nothing follows /breathe, so no trailing space — the send button
        // submits instead of sending "/breathe ".
        assertEquals("/breathe", ChatCommands.completionFor(catalog[2]))
    }
}

File: desktop/ui/chat_commands.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  ACTION_KEYS,
  ASIDE,
  BLOCKED,
  CHOOSE,
  DESKTOP_CAN_PLAY,
  DESKTOP_SCREENS,
  HELP,
  INCOMPLETE,
  OFFER,
  OPEN,
  SEND,
  TARA_HERE,
  TARA_PRIVATE,
  TARA_SHARED,
  TASK,
  candidateLabel,
  completionFor,
  labelFor,
  offerAffordance,
  pickerRows,
  planFor,
  splitAuthor,
  taraDraft,
  withChoices,
} from "./chat_commands.mjs";

// The bridge shapes, as serde delivers them.
const plain = { kind: "plain" };
const task = (text, assignees = []) => ({ kind: "task", text, assignees });
const offer = (action, targets) => ({ kind: "offer_to", action, targets });
const mention = (handle) => ({ handle, start: 0, end: handle.length + 1 });
const resolved = (handle, npub) => ({ handle, npub, candidates: [] });
const ambiguous = (handle) => ({
  handle,
  npub: null,
  candidates: [
    { npub: "npub1a", alias: "ana" },
    { npub: "npub1b", alias: "ana" },
  ],
});

test("ordinary text is sent as typed", () => {
  assert.deepEqual(planFor(plain), { action: SEND });
});

test("a command this build has never heard of is reported, never swallowed", () => {
  // A message silently eaten as an unknown command is worse than one that
  // arrives confusing — the sender thinks they sent it.
  const p = planFor({ kind: "unknown", name: "frobnicate" });
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /no \/frobnicate/);
});

test("a grammar variant newer than this file falls back to sending", () => {
  // Release skew between the Rust grammar and this window, not a bug. Sending
  // the text is the safe outcome.
  assert.deepEqual(planFor({ kind: "something_new_entirely" }), { action: SEND });
  assert.deepEqual(planFor(null), { action: SEND });
  assert.deepEqual(planFor({}), { action: SEND });
});

test("/pay still goes out as text", () => {
  // Recognised so it is not called a typo; the chip and the peer's own client
  // do the rest, exactly as before this feature existed.
  assert.deepEqual(planFor({ kind: "pay" }), { action: SEND });
});

test("/help opens the list", () => {
  assert.deepEqual(planFor({ kind: "help" }), { action: HELP });
});

// ── Tara asides ──────────────────────────────────────────────────────────────

test("an aside routes to tara, never to the peer", () => {
  const p = planFor({ kind: "ask_tara", text: "i keep putting this off" });
  assert.equal(p.action, ASIDE);
  assert.equal(p.text, "i keep putting this off");
});

test("an addressed but empty aside prompts instead of sending an empty turn", () => {
  const p = planFor({ kind: "ask_tara", text: "  " });
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /only you will see it/);
});

test("the composer knows the audience from the moment tara is addressed", () => {
  // Eager on purpose, and now in two directions: `/tara` never leaves the device
  // and `@tara` reaches the other person. Either label being wrong is how
  // somebody publishes a private thought, or asks a question the peer never sees.
  assert.equal(taraDraft("@tara "), TARA_SHARED);
  assert.equal(taraDraft("@tara"), TARA_SHARED);
  assert.equal(taraDraft("@Tara what about"), TARA_SHARED);
  assert.equal(taraDraft("  @tara hello"), TARA_SHARED);
  assert.equal(taraDraft("/tara hello"), TARA_PRIVATE);
  assert.equal(taraDraft("/tara"), TARA_PRIVATE);
});

test("a person whose handle starts with tara is not tara", () => {
  assert.equal(taraDraft("@taranjeet are you around"), null);
  assert.equal(taraDraft("hello @tara"), null);
  assert.equal(taraDraft(""), null);
  assert.equal(taraDraft(undefined), null);
});

// ── Tara in the room (`@tara …`) ──────────────────────────────────────────────

test("shared and private tara are different plans", () => {
  // Same words, two audiences: one reaches `tara_in_chat` (two messages into the
  // thread), the other `tara_aside` (nothing sent).
  const shared = planFor({ kind: "tara_here", text: "what film should we watch" });
  assert.equal(shared.action, TARA_HERE);
  assert.equal(shared.text, "what film should we watch");
  assert.equal(planFor({ kind: "ask_tara", text: "what film should we watch" }).action, ASIDE);
});

test("an empty shared address says the answer will be seen by both", () => {
  const p = planFor({ kind: "tara_here", text: " " });
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /you'll both see/);
});

// ── Tasks ────────────────────────────────────────────────────────────────────

test("a task with nobody named is a note to self", () => {
  const p = planFor(task("water the plants"));
  assert.equal(p.action, TASK);
  assert.equal(p.text, "water the plants");
  assert.equal(p.peer, null);
});

test("a task naming a known contact carries their npub", () => {
  const p = planFor(task("get some work done", [mention("ana")]), {
    mentions: [resolved("ana", "npub1ana")],
  });
  assert.equal(p.action, TASK);
  assert.equal(p.peer, "npub1ana");
});

test("an empty task asks what needs doing", () => {
  const p = planFor(task("   "));
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /What needs doing/);
});

test("a handle that is nobody's is reported rather than guessed at", () => {
  const p = planFor(task("do it", [mention("bina")]), {
    mentions: [{ handle: "bina", npub: null, candidates: [] }],
  });
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /@bina isn't in your contacts/);
});

test("two contacts answering to one handle becomes a question, not a coin flip", () => {
  // A handle is a self-declared alias; anyone may publish any name. Picking one
  // is how a private message reaches the wrong person.
  const p = planFor(task("do it", [mention("ana")]), {
    mentions: [
      { handle: "ana", npub: null, candidates: [{ npub: "npub1a" }, { npub: "npub1b" }] },
    ],
  });
  assert.equal(p.action, CHOOSE);
  assert.equal(p.handle, "ana");
  assert.equal(p.candidates.length, 2);
});

test("choosing one of two lets the same draft through", () => {
  // The fix for the dead end: the old plan said "pick which one" and gave
  // nothing to pick, so the command could never be completed at all.
  const mentions = withChoices([ambiguous("ana")], { ana: "npub1b" });
  const p = planFor(task("do it", [mention("ana")]), { mentions });
  assert.equal(p.action, TASK);
  assert.equal(p.peer, "npub1b");
});

test("a choice that names nobody on the list is ignored", () => {
  // A pin left over from an earlier draft must not silently retarget a message —
  // that is the failure the ambiguity exists to prevent, arriving by another door.
  const stale = withChoices([ambiguous("ana")], { ana: "npub1someone-else" });
  assert.equal(stale[0].npub, null);
  assert.equal(planFor(task("do it", [mention("ana")]), { mentions: stale }).action, CHOOSE);
});

test("a chooser row carries the key because both may share the name", () => {
  // The usual reason a handle is ambiguous is that both people chose the same
  // name, so a row showing only the name offers two identical choices.
  const a = candidateLabel("ana", "npub1aaaaaaaaaaaaaaaaaaaa");
  const b = candidateLabel("ana", "npub1bbbbbbbbbbbbbbbbbbbb");
  assert.notEqual(a, b);
  assert.match(a, /ana/);
});

test("an ambiguous handle in an offer asks too, rather than picking one", () => {
  const p = planFor(offer("breathe", [mention("ana")]), { mentions: [ambiguous("ana")] });
  assert.equal(p.action, CHOOSE);
});

// ── Offers ───────────────────────────────────────────────────────────────────

test("an offer to a known comrade carries the action and the npub", () => {
  const p = planFor(offer("breathe", [mention("ana")]), {
    mentions: [resolved("ana", "npub1ana")],
  });
  assert.equal(p.action, OFFER);
  assert.equal(p.appAction, "breathe");
  assert.deepEqual(p.peers, ["npub1ana"]);
});

test("an offer with nobody named asks who it is for", () => {
  const p = planFor(offer("breathe", []));
  assert.equal(p.action, INCOMPLETE);
  assert.match(p.message, /Name the comrade/);
});

test("an offer to an unresolvable handle does not silently go nowhere", () => {
  const p = planFor(offer("breathe", [mention("ghost")]), {
    mentions: [{ handle: "ghost", npub: null, candidates: [] }],
  });
  assert.equal(p.action, INCOMPLETE);
});

// ── Screens this window does and does not have ───────────────────────────────

test("a screen desktop has opens; one it lacks says so instead of doing nothing", () => {
  // `docs/FRONTEND_STRATEGY.md`'s divergence table: journal, Tara and the
  // breathing screen are Android-only today. A dead button is the worse answer.
  assert.deepEqual(planFor({ kind: "open", action: "focus" }), {
    action: OPEN,
    appAction: "focus",
  });
  const blocked = planFor({ kind: "open", action: "journal" });
  assert.equal(blocked.action, BLOCKED);
  assert.match(blocked.message, /not here yet/);
  assert.match(blocked.message, /journal/i);
});

test("every desktop screen in the set actually opens", () => {
  for (const key of DESKTOP_SCREENS) {
    assert.equal(planFor({ kind: "open", action: key }).action, OPEN, key);
  }
});

test("play is refused honestly while there is no player in this window", () => {
  // `docs/TOGETHER.md` §9: the commands and the decision module are here, the
  // <video> element and the file picker are not. When that lands, flip
  // DESKTOP_CAN_PLAY and this test's other branch takes over.
  const p = planFor({ kind: "play", query: "Kun Faya Kun", service: "spotify" });
  if (DESKTOP_CAN_PLAY) {
    assert.equal(p.action, "play");
    assert.equal(p.query, "Kun Faya Kun");
  } else {
    assert.equal(p.action, BLOCKED);
    assert.match(p.message, /no player here yet/);
  }
});

// ── The / picker ─────────────────────────────────────────────────────────────

const CATALOG = [
  { name: "play", aliases: ["listen", "watch"], argument: "<song>", takes_mention: false },
  { name: "task", aliases: ["todo"], argument: "<what>", takes_mention: true },
  { name: "breathe", aliases: ["breath"], argument: "", takes_mention: false },
];

test("the picker matches aliases as well as names", () => {
  // An alias that works on submit but is invisible while typing reads as not
  // existing.
  const rows = pickerRows("/lis", CATALOG);
  assert.equal(rows.length, 1);
  assert.equal(rows[0].name, "play");
});

test("the picker is closed rather than empty for ordinary text", () => {
  assert.equal(pickerRows("hello", CATALOG), null);
  assert.equal(pickerRows("20/80 split", CATALOG), null);
  assert.equal(pickerRows("/nomatch", CATALOG), null);
  assert.equal(pickerRows(undefined, CATALOG), null);
});

test("the picker closes once the command has been chosen", () => {
  // A space means what follows is an argument, not a name.
  assert.equal(pickerRows("/task ship it", CATALOG), null);
});

test("a bare slash offers everything", () => {
  assert.equal(pickerRows("/", CATALOG).length, 3);
});

test("completion leaves the caret where the next word goes, or submits", () => {
  assert.equal(completionFor(CATALOG[0]), "/play ");
  assert.equal(completionFor(CATALOG[1]), "/task ");
  // Nothing follows /breathe, so no trailing space — Enter submits instead of
  // sending "/breathe ".
  assert.equal(completionFor(CATALOG[2]), "/breathe");
  assert.equal(completionFor(null), "");
});

// ── Received offers ──────────────────────────────────────────────────────────

test("every action key the bridge can send has a real label", () => {
  // The drift this catches: `labelFor` falls back to the raw key, so a
  // mismatched spelling produced "breathe is on the phone app, not here yet"
  // and nothing failed. Assert the exact strings, not a loose /breath/i.
  for (const key of ACTION_KEYS) {
    assert.notEqual(labelFor(key), key, `${key} fell through to the raw key`);
  }
  assert.equal(labelFor("breathe"), "Taking a deep breath");
  assert.equal(labelFor("read"), "The reader");
});

test("every action key is either a desktop screen or honestly refused", () => {
  for (const key of ACTION_KEYS) {
    const plan = planFor({ kind: "open", action: key });
    if (DESKTOP_SCREENS.has(key)) {
      assert.equal(plan.action, OPEN, key);
    } else {
      assert.equal(plan.action, BLOCKED, key);
      // The sentence must name the thing, not echo a wire key at the user.
      assert.ok(!plan.message.startsWith(`${key} is`), `${key} leaked its wire key`);
    }
  }
});

test("a received offer is shown even when this window lacks the screen", () => {
  // The sentence is the point; hiding a comrade's message because our build has
  // no screen would be the wrong trade.
  const here = offerAffordance("focus");
  assert.equal(here.actionable, true);
  const elsewhere = offerAffordance("breathe");
  assert.equal(elsewhere.actionable, false);
  assert.equal(elsewhere.label, "Taking a deep breath");
  assert.equal(offerAffordance(null), null);
});

test("a shared Tara answer is split the same way core splits it", () => {
  // The mirror of `comrade_ui::split_author`, needed because a live-arriving DM
  // reaches this frontend as the raw wire body rather than through `MessageDto`.
  assert.deepEqual(splitAuthor("Tara: one thing at a time"), {
    author: "tara",
    content: "one thing at a time",
  });
  assert.deepEqual(splitAuthor("what do you think"), {
    author: "human",
    content: "what do you think",
  });
  // A prefix, not a substring: "Tara" is an ordinary word in an ordinary
  // sentence, and attributing this to her would put words in her mouth.
  assert.equal(splitAuthor("Tara said something like that").author, "human");
  assert.equal(splitAuthor("ask Tara: what now").author, "human");
  // Anyone can type the marker. This asserts the limitation rather than hiding
  // it: a match is the sender's claim, never proof — AUDIT.md Q17.
  assert.equal(splitAuthor("Tara: I made this up").author, "tara");
  assert.deepEqual(splitAuthor(null), { author: "human", content: "" });
});
